### ISSUE TITLE
[Feature] Support Snowflake China host OAuth connection snowflakecomputing.cn

### ISSUE THREAD

**Issue by AndrewCanva** [dbt-labs/dbt-snowflake#1024]
### Is this your first time submitting a feature request?

- [X] I have read the [expectations for open source contributors](https://docs.getdbt.com/docs/contributing/oss-expectations)
- [X] I have searched the existing issues, and I could not find an existing issue for this feature
- [X] I am requesting a straightforward extension of existing dbt-snowflake functionality, rather than a Big Idea better suited to a discussion

### Describe the feature

Snowflake has initiated the private preview of their product in China, with plans to transition into a public preview in June 2024. Presently, connections to Snowflake China are not possible through the standard snowflakecomputing.com host; instead, the regional host snowflakecomputing.cn must be used.

### Describe alternatives you've considered

N/A

### Who will this benefit?

dbt and snowflake users in China

### Are you interested in contributing this feature?

yes

### Anything else?

N/A

**Comment by jtcohen6** [comment]
Hey @AndrewCanva, thanks for opening!

Do you happen to know more details (or could link us to documentation) about Snowflake China?

dbt-snowflake connects to Snowflake using [`snowflake-connector-python`](https://github.com/snowflakedb/snowflake-connector-python), by passing in the [account identifier](https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-api#usage-notes-for-the-account-parameter-for-the-connect-method). However `snowflake-python-connector` wants to receive region/organization/account identifiers for Snowflake China, that's how we'll want to provide it as well.

**Comment by AndrewCanva** [comment]
> Hey @AndrewCanva, thanks for opening!
> 
> Do you happen to know more details (or could link us to documentation) about Snowflake China?
> 
> dbt-snowflake connects to Snowflake using [`snowflake-connector-python`](https://github.com/snowflakedb/snowflake-connector-python), by passing in the [account identifier](https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-api#usage-notes-for-the-account-parameter-for-the-connect-method). However `snowflake-python-connector` wants to receive region/organization/account identifiers for Snowflake China, that's how we'll want to provide it as well.

@jtcohen6 the error was gone after I added `host` parameter in the connection profile because I used snowflake/externalbrowser authenticator, but I guess it would cause errors if oAuth authenticator method is configured re https://github.com/dbt-labs/dbt-snowflake/blob/5459fbf0aa1bb40e800b89200442fcab47504157/dbt/adapters/snowflake/connections.py#L57

**Comment by jtcohen6** [comment]
@AndrewCanva Makes sense. I'm going to keep this one on the backlog for now, and we can pick it up when someone needs it. Good to have it searchable in the meantime.
